app/notebook/server/NotebookManager.scala

```diff
@@ -0,0 +1,144 @@
+    val nb = Notebook(
+      Some(new Metadata(getName(fpath),
+        customLocalRepo = customLocalRepo,
+        customRepos = customRepos,
+        customDeps = customDeps,
+        customImports = customImports,
+        customSparkConf = customSparkConf)),
+      Some(Nil),
+      None,
+      None,
+      None
+    )
+    save(fpath, nb, overwrite = false)
+    fpath
+  }
+
+  def copyNotebook(nbPath: String) = {
+    val nbData = getNotebook(nbPath)
+    nbData.map { nb =>
+      val newPath = incrementFileName(nb._4.dropRight(extension.length))
+      val newName = getName(newPath)
+      val oldNB = NBSerializer.read(nb._3)
+      save(newPath, Notebook(oldNB.metadata.map(_.copy(name = newName)), oldNB.cells, oldNB.worksheets, oldNB.autosaved, None), false)
+      newPath
+    } getOrElse newNotebook()
+  }
+
+  def getNotebook(path: String) = {
+    Logger.info(s"getNotebook at path $path")
+    for (notebook <- load(path)) yield {
+      val data = FileUtils.readFileToString(notebookFile(path))
+      val df = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
+      val last_mtime = df.format(new Date(notebookFile(path).lastModified()))
+      (last_mtime, notebook.name, data, path)
+    }
+  }
+
+  def deleteNotebook(path: String) = {
+    Logger.info(s"deleteNotebook at path $path")
+    val file = notebookFile(path)
+    if (file.exists()) {
+      file.delete()
+    }
+  }
+
+  def rename(path: String, newpath: String) = {
+    Logger.info(s"rename from path $path to $newpath")
+    val newname = getName(newpath)
+    val oldfile = notebookFile(path)
+    Logger.debug(s"rename from path $path to $newpath: old file is ${oldfile.getAbsolutePath}")
+    load(path).foreach { notebook =>
+      val nb = if (notebook.name != newname) {
+        val meta = notebook.metadata.map(_.copy(name = newname)).orElse(Some(new Metadata(newname)))
+        notebook.copy(metadata = meta)
+      } else {
+        notebook
+      }
+      val newfile = notebookFile(newpath)
+      Logger.debug(s"rename from path $path to $newpath: new file is ${newfile.getAbsolutePath}")
+      oldfile.renameTo(newfile)
+      FileUtils.writeStringToFile(newfile, NBSerializer.write(nb))
+    }
+    (newname, newpath)
+  }
+
+  def save(path: String, notebook: Notebook, overwrite: Boolean) = {
+    Logger.info(s"save at path $path")
+    val file = notebookFile(path)
+    if (!overwrite && file.exists()) {
+      throw new NotebookExistsException("Notebook " + path + " already exists.")
+    }
+    FileUtils.writeStringToFile(file, NBSerializer.write(notebook))
+    val nb = load(path)
+    (nb.get.metadata.get.name, path)
+  }
+
+  def load(path: String): Option[Notebook] = {
+    Logger.info(s"Loading notebook at path $path")
+    val file = notebookFile(path)
+    if (file.exists())
+      Some(NBSerializer.read(FileUtils.readFileToString(file)))
+    else None
+  }
+
+
+}
+
+class NotebookExistsException(message: String) extends IOException(message)
```

Comment:
interesting, is there some reasons you like that? (I'm not against it, just curious :smile: )